includes/sanitizers/class-amp-core-theme-sanitizer.php

```diff
@@ -1522,4 +1569,446 @@ public function add_twentyfourteen_search() {
 		$search_toggle_div->setAttribute( AMP_DOM_Utils::AMP_BIND_DATA_ATTR_PREFIX . 'class', "$hidden_state_id ? 'search-toggle' : 'search-toggle active'" );
 		$search_container->setAttribute( AMP_DOM_Utils::AMP_BIND_DATA_ATTR_PREFIX . 'class', "$hidden_state_id ? 'search-box-wrapper hide' : 'search-box-wrapper'" );
 	}
+
+	/**
+	 * Wrap a modal node tree in an <amp-lightbox> element.
+	 *
+	 * @param array $args {
+	 *     Associative array of arguments.
+	 *
+	 *     @type string   $modal_id            ID to use for the modal and its associated buttons.
+	 *     @type string   $modal_content_xpath XPath to query the contents of the modal.
+	 *     @type string[] $open_button_xpath   Array of XPaths to query the buttons that open the modal.
+	 *     @type string[] $close_button_xpath  Array of XPaths to query the buttons that close the modal. These should be contained within the modal.
+	 *     @type string   $animate_in          Optional. What animation to use for showing the modal. Valid options are: 'fade-in', 'fly-in-bottom', 'fly-in-top'. Defaults to 'fade-in'.
+	 *     @type bool     $scrollable          Optional. Whether the inner content of the modal should be scrollable. Defaults to true.
+	 * }
+	 */
+	public function wrap_modal_in_lightbox( $args = [] ) {
+		if ( ! isset( $args['modal_id'], $args['modal_content_xpath'], $args['open_button_xpath'], $args['close_button_xpath'] ) ) {
+			return;
+		}
+
+		$modal_id           = $args['modal_id'];
+		$modal_content_node = $this->xpath->query( $args['modal_content_xpath'] )->item( 0 );
+
+		if ( ! is_string( $modal_id ) || ! $modal_content_node instanceof DOMElement ) {
+			return;
+		}
+
+		$body_id = $this->get_element_id( $this->get_body_node(), 'body' );
+
+		$open_xpaths  = isset( $args['open_button_xpath'] ) ? $args['open_button_xpath'] : [];
+		$close_xpaths = isset( $args['close_button_xpath'] ) ? $args['close_button_xpath'] : [];
+
+		$buttons = [
+			"{$modal_id}.open"  => $open_xpaths,
+			"{$body_id}.toggleClass(class=showing-modal,force=true)" => $open_xpaths,
+			"{$modal_id}.close" => $close_xpaths,
+			"{$body_id}.toggleClass(class=showing-modal,force=false)" => $close_xpaths,
+		];
+
+		foreach ( $buttons as $action => $button_xpaths ) {
+			foreach ( $button_xpaths as $button_xpath ) {
+				foreach ( $this->xpath->query( $button_xpath ) as $node ) {
```

Comment:
This `O(n^3)` looks scary. Would be good to add a comment explaining the expectation for the total number of iterations in the innermost loop.

Reply:
For the `aria-expanding` code, I need to refactor back into using `<amp-state>`, so this will be gone again.

Reply:
I made this clearer in https://github.com/ampproject/amp-wp/pull/3403/commits/abe0b0691a0a5ceabbf305602c7de02a386d4097